feat: perun_user_microservice

PeterBolha opened the pull request.

melanger reviewed (changes requested):
Use conventional commit message, e.g. "feat: get perun user microservice", and set PR status to Draft

PeterBolha commented:
> Use conventional commit message, e.g. "feat: get perun user microservice", and set PR status to Draft

Initially, in AdaptersManager we used a workflow where commits were constantly squashed with each new push so there was always only a commit like "feat: get perun user microservice". Then we switched to keeping temporary history which would be squashed upon merging into main. To my understanding, the conventional commits with standardized labels like _feat_, _fix_ etc. were reserved for squashed, cohesive groups of changes ready to be merged into main. So there could be a group of commits labeled like:

A - Minor change of something
B - Minor tweak of something else
C - Addition of attribute

and if commits A,B and C completed some bit of functionality, they would be squashed into something like:
A - feat: implemented something

Perhaps I misunderstood the workflow earlier. I don't have a problem with labeling everything in the standardized format, I just want to make sure we're all on the same page beforehand. Should the above scenario look like:

A - fix: minor change of something
B - fix: minor tweak of something else
C - feat: addition of attribute

Or should I perhaps keep it to "A - feat: implemented something" at all times and squash before each push?

melanger commented:
> > Use conventional commit message, e.g. "feat: get perun user microservice", and set PR status to Draft
> 
> Initially, in AdaptersManager we used a workflow where commits were constantly squashed with each new push so there was always only a commit like "feat: get perun user microservice". Then we switched to keeping temporary history which would be squashed upon merging into main. To my understanding, the conventional commits with standardized labels like _feat_, _fix_ etc. were reserved for squashed, cohesive groups of changes ready to be merged into main. So there could be a group of commits labeled like:
> 
> A - Minor change of something B - Minor tweak of something else C - Addition of attribute
> 
> and if commits A,B and C completed some bit of functionality, they would be squashed into something like: A - feat: implemented something
> 
> Perhaps I misunderstood the workflow earlier. I don't have a problem with labeling everything in the standardized format, I just want to make sure we're all on the same page beforehand. Should the above scenario look like:
> 
> A - fix: minor change of something B - fix: minor tweak of something else C - feat: addition of attribute
> 
> Or should I perhaps keep it to "A - feat: implemented something" at all times and squash before each push?

If A,B,C are just modifications of a single resulting commit (which is probably this case), then I'm fine with the first approach, leave out the prefixes and squash before merge.
If, for example, A+B produce a "feat:" commit but C is a different "fix:", then of course the first approach does not work. That's why I (personally) squash continuously when needed and I create a PR once everything seems ready.

In this case, while the PR is a Draft, you can leave out the prefixes.

melanger commented on microservices/perun_user_microservice.py:
```diff
@@ -0,0 +1,62 @@
+        self.__internal_extsource_attribute = config[
+            "internal_extsource_attribute"]
+        self.__proxy_extsource_name = config["proxy_extsource_name"]
+        self.__allowed_requesters = config.get("allowed_requesters")
```
`config.get("allowed_requesters")` is the same as `config["allowed_requesters"]`. If you want to have an optional parameter, you have to supply a default value, e.g. `config.get("allowed_requesters", [])` or `config.get("allowed_requesters", None)`

PeterBolha replied:
I think `None` is the default return value for `get()` method on python dict, therefore `config.get("key"`) and `config.get("key", None)` should be equivalent. In perun identity beta, `None` was used as the default value, so I shortened it to `config.get("allowed_requesters")`, but I can keep it explicit if that seems better. Perhaps it's more easily understandable that way.

melanger replied:
You're right, `None` is the default default, so `.get()` is different from `[]`.

You might still want to use an empty list instead of None, because then you don't have to branch.

PeterBolha replied:
Yes, that's a good idea. I'll add `[]` default return values.

melanger commented on microservices/perun_user_microservice.py:
```diff
@@ -0,0 +1,62 @@
+                        f"found in Perun. And redirect link to registration "
+                        f"page is missing in the config file.")
+
+                # TODO make sure this is how redirect works
```
Take a look at e.g. https://github.com/CESNET/satosa-microservice-webauthn/blob/c9ae290900cdc128c46d8af8788b1ed251cba93e/webauthn_microservice.py#L89

PeterBolha commented on microservices/perun_user_microservice.py:
```diff
@@ -0,0 +1,62 @@
+                data.attributes["id"] = user.id
+
+        except KeyError as missing_key:
+            raise SATOSAError(f"Missing key: {missing_key} in config file.")
```
I think my mistake was rather naming error improperly. I put this except block there because of the line `name = data["auth_info"]["issuer"]`. I'm not sure whether `data` can actually have missing keys, but I put it there just in case. I could rename the error or get rid of the except block altogether.

melanger replied:
[InternalData](https://github.com/IdentityPython/SATOSA/blob/1de6a360d52369e01a40d7d0400af20735d7e899/src/satosa/internal.py#L114) is a call and [AuthenticationInformation](https://github.com/IdentityPython/SATOSA/blob/1de6a360d52369e01a40d7d0400af20735d7e899/src/satosa/internal.py#L82) is a class, the keys cannot be missing, so you don't have to test for this.

PeterBolha replied:
Alright that makes sense, I removed the checking.

xpavlic commented on utils/jws.py:
```diff
@@ -0,0 +1,59 @@
+        return jwk_set
+
+    # TODO find out if we actually need api signing key
+    def _get_api_jwk(self):
```
remove

xpavlic commented on utils/jws.py:
```diff
@@ -0,0 +1,59 @@
+            self._CONFIG["perun_storage"]["signing"]["key_id"])
+
+    def _get_jwt(self, data, jwk_key):
+        current_timestamp = int(time.time())
```
```suggestion
```

xpavlic commented on utils/jws.py:
```diff
@@ -0,0 +1,59 @@
+from satosa.response import Redirect
+
+
+class RedirectService:
```
add public static method generate_noonce 

letters = string.ascii_lowercase
actual_time = str(int(time.time()))
rand = random.SystemRandom()
return actual_time + ''.join(rand.choice(letters) for _ in range(54))

xpavlic commented on microservices/perun_user_microservice.py:
```diff
@@ -0,0 +1,58 @@
+
+                return Redirect(self.__registration_page_url)
+
+        return super().process(context, data)
```
You have to register endpoint, and after redirect it will get back into registered method and do another request to check if registration was suceessfull. You can find inspiration in.  [https://github.com/CESNET/satosa-microservice-webauthn/blob/c9ae290900cdc128c46d8af8788b1ed251cba93e/webauthn_microservice.py#L12](satosa-microservice-webauthn). If you need more help with that, let me know.

PeterBolha replied:
I'm not sure I understand the process of endpoint registration. In `register_endpoint` method, `_handle_webauthn_response` is called with no given parameter even though this method takes is context parameter which it further uses. It also uses user_id, signing_key and api_url from config. Will we use these attributes as well?

melanger replied:
The endpoint is called after user is redirected back from the external page to SATOSA. If you need any data there, you have to save them into session and recover them after the user gets back. See https://github.com/CESNET/satosa-microservice-webauthn/blob/master/webauthn_microservice.py - there it is called "saved_state".

PeterBolha replied:
So to my understanding, we want to obtain perun user based on their credentials. We sign this request, send it to an external page while saving current user's data in a session. We register an endpoint where the reply from the external website arrives. If the request is not allowed or the reply is somehow invalid, we raise an exception. So the registering of endpoint is not calling the method in any way, it just tells the external website where to reply and the context is supplied from our session?

melanger replied:
> So to my understanding, we want to obtain perun user based on their credentials. We sign this request, send it to an external page while saving current user's data in a session. We register an endpoint where the reply from the external website arrives. If the request is not allowed or the reply is somehow invalid, we raise an exception.

Yes.

> So the registering of endpoint is not calling the method in any way, it just tells the external website where to reply and the context is supplied from our session?

It does not tell the external website, registration of endpoint just creates a URL on the proxy to which the external website (or our separate module) can redirect back to ("reply to"). The registration method is only called once when SATOSA is starting. The method which is registered on the endpoint is called whenever the external website/module redirects user to it, so it needs to read data from the external source and also extract user data (context) from session.

melanger replied:
Check out https://github.com/CESNET/satosa-microservice-webauthn/blob/master/webauthn_microservice.py - the process method makes a redirect to external module, the external module redirects back to the endpoint registered in `register_endpoints`, so the method `_handle_webauthn_response` is called, which gets data from external module (here by backchannel) and extracts user context from session (`saved_state = context.state[self.name]` etc.)

melanger commented on utils/jws.py:
```diff
@@ -0,0 +1,59 @@
+
+    # TODO add source class and method to obtain default config from
+    def __init__(self, config=None):
+        self._CONFIG = config
```
please use e.g. ECDSA as the default, not RSA (which is considered not as secure and requires much longer keys)

xpavlic commented on microservices/perun_user_microservice.py:
```diff
@@ -0,0 +1,89 @@
+        self.__adapters_manager = AdaptersManager()
+        self._endpoint = "/process"
+
+    def _handle_webauthn_response(self, context):
```
```suggestion
    def _handle_registration_response(self, context):
```

xpavlic commented on microservices/perun_user_microservice.py:
```diff
@@ -0,0 +1,89 @@
+                   "nonce": internal_response['nonce'],
+                   "time": str(int(time.time()))}
+        message_json = json.dumps(message)
+        jws = JWS(message_json, alg=self.signing_key.alg).sign_compact(
```
Use method for signing in utils.

xpavlic reviewed (changes requested):
Rebase your branch, there are few updates in requirements.

xpavlic commented on microservices/perun_user_microservice.py:
```diff
@@ -0,0 +1,94 @@
+                data.attributes[self.__internal_extsource_attribute] = [
+                    self.__proxy_extsource_name
+                ]
+                # TODO add configurable attribute name for user id
```
attr name will be in global cfg. Example is above.

xpavlic commented on microservices/perun_user_microservice.py:
```diff
@@ -0,0 +1,94 @@
+
+            user = self.__adapters_manager.get_perun_user(name, logins)
+            if user:
+                data.subject_id = user.id
```
According to perun_identity_beta, data.subject_id should be login which is user attribute and different of user id which we want in attributes.

PeterBolha replied:
So if I understand it correctly - in `perun_identity_beta` the `login` in `data.subject_id = login` is the entire user extracted by the rpc request?

xpavlic replied:
Login is value of configurable attribute of user from perun not User object.

xpavlic commented on microservices/perun_user_microservice.py:
```diff
@@ -0,0 +1,94 @@
+        if response_dict["result"] != "okay" or not hmac.compare_digest(
+                response_dict["nonce"], internal_response["nonce"]):
+            raise Exception("Authentication was unsuccessful.")
+        if "authn_context_class_ref" in context.state:
+            internal_response["auth_info"]["auth_class_ref"] = context.state[
+                "authn_context_class_ref"]
```
We need to perform same functionality with registered user as in process to save user data.
```suggestion
```

xpavlic replied:
probably you can call return process(context, internal_response)

xpavlic commented on utils/utils.py:
```diff
@@ -0,0 +1,49 @@
+        return token.serialize()
+
+    @staticmethod
+    def sign_data(data: dict[str, str], config: dict[str, str] = None) -> str:
```
I would rather pass parameters like keystore etc. instead of config

PeterBolha commented on microservices/perun_user_microservice.py:
```diff
@@ -0,0 +1,94 @@
+
+
+class PerunUser(ResponseMicroService):
+    def __init__(self, config, *args, **kwargs):
```
Maybe I'd rename `adapters_manager`  to `adapters_manager_cfg` to be consistent with other config names

xpavlic replied:
In my opinion adapters_manager is ok because there are only configs in this file so it is not needed to write cfg in every option name.

PeterBolha replied:
Alright, that makes sense. I'll keep it like that and add this global config.

PeterBolha commented on microservices/perun_user_microservice.py:
```diff
@@ -0,0 +1,94 @@
+                # TODO add configurable attribute name for user id
+                data.attributes["id"] = user.id
+
+        except ApiException as api_exception:
```
I think RPC can do the same thing. Shouldn't we handle it in a same way as the ApiException? From logical standpoint, it's the same scenario.

xpavlic replied:
I will add throwing exceptions from LDAP Adapter with changes in attributes.

xpavlic commented on config_templates/global.yaml:
```diff
@@ -0,0 +1,35 @@
+     priority: 1
+     username: cn=admin,dc=muni,dc=cz
+     base_dn: dc=muni,dc=cz
+     password: mypassword
```
add `rp_id_attribute: urn:perun:facility:attribute-def:def:rpIdentifier` as new configuration option.

xpavlic commented on config_templates/global.yaml:
```diff
@@ -0,0 +1,35 @@
+     #ApiKeyAuth
+     api_key: your_api_key
+     #BearerAuth
+     access_token: your_bearer_token
```
add `rp_id_attribute: urn:perun:facility:attribute-def:def:rpIdentifier` as new configuration option.

xpavlic commented on microservices/perun_user_microservice.py:
```diff
@@ -0,0 +1,118 @@
+        logins = data.attributes[self.__internal_login_attribute]
+
+        user = self.__adapters_manager.get_perun_user(name, logins)
+        if user is None:
```
```suggestion
        if not user:
```

xpavlic commented on microservices/perun_user_microservice.py:
```diff
@@ -0,0 +1,118 @@
+                f"found in Perun. And redirect link to registration "
+                f"page is missing in the config file.")
+
+        nonce = {"nonce": Utils.generate_nonce(),
+                 "time": str(int(time.time()))}
+        signed_nonce = self.sign_data(nonce)
+        data['nonce'] = nonce
+        context.state[self.name] = data.to_dict()
+        return Redirect(f"{self.__registration_page_url}/{signed_nonce}/")
```
```suggestion
        nonce = Utils.generate_nonce()
        request_data = {"nonce": nonce,
                        "time": str(int(time.time()))}
        signed_request_data = self.sign_data(request_data)
        data['nonce'] = nonce
        context.state[self.name] = data.to_dict()
        return Redirect(f"{self.__registration_page_url}/{signed_request_data}")
```

xpavlic commented on utils/utils.py:
```diff
@@ -0,0 +1,45 @@
+    def sign_data(data: dict[str, str], keystore: str, key_id: str,
+                  token_alg: str) -> str:
+        token_signing_key = Utils.__get_signing_jwk(keystore, key_id)
+        signed_data = Utils.__get_jwt(data, token_signing_key, token_alg)
+
+        return signed_data
```
```suggestion
        return Utils.__get_jwt(data, token_signing_key, token_alg)
```

xpavlic commented on microservices/perun_user_microservice.py:
```diff
@@ -0,0 +1,118 @@
+        message = {"nonce": internal_response['nonce'],
+                   "time": str(int(time.time()))}
+
+        keystore = self.__signing_cfg["keystore"]
+        key_id = self.__signing_cfg["key_id"]
+        token_alg = self.__signing_cfg["token_alg"]
+
+        signed_message = Utils.sign_data(message, keystore, key_id, token_alg)
+        request = f"{self.__api_url}/{signed_message}"
```
```suggestion
        data = {"nonce": internal_response['nonce'],
                "time": str(int(time.time()))}

        signed_data = self.sign_data(data)
        request = f"{self.__api_url}/{signed_data}"
```

xpavlic commented on microservices/perun_user_microservice.py:
```diff
@@ -0,0 +1,118 @@
+
+        if response_dict["result"] != "okay" or not hmac.compare_digest(
+                response_dict["nonce"], internal_response["nonce"]):
+            raise Exception("Authentication was unsuccessful.")
```
```suggestion
            raise Exception("Registration was unsuccessful.")
```

xpavlic commented:
After https://github.com/CESNET/perun.connector/pulls will be merged, update perun.connector version.

xpavlic commented on satosacontrib/perun/micro_services/perun_user_microservice.py:
```diff
@@ -0,0 +1,140 @@
+            config["global_cfg_filepath"]
+        )
+
+        self.__perun_user_id_attr = global_config["perun_user_id_attribute"]
+        self.__perun_login_attribute = global_config["perun_login_attribute"]
```
Does not match with global config in templates.

xpavlic commented on satosacontrib/perun/utils/Utils.py:
```diff
@@ -0,0 +1,124 @@
+        return Utils.__get_jwt(data, token_signing_key, token_alg)
+
+    @staticmethod
+    def secure_redirect_with_nonce(
```
In my opinion, we should pass request data as parameter. In some cases we probably want send another informations not only nonce and time. In this microservice you could just pass empty dict and in this method just add nonce and time to it.

xpavlic commented on requirements.txt:
```diff
@@ -1 +1,8 @@
-.
\ No newline at end of file
+SATOSA~=8.1.0
+pysaml2~=7.1.2
+requests~=2.27.1
+perun.connector~=2.1.0
+pycurl~=7.45.1
+PyYAML~=6.0
+jwcrypto==1.3.1
+pytest~=7.1.2
```
the . is correct content. We use setup.py to install dependencies.

melanger replied:
test dependencies have to be separated, something like:
```python
install_requires = [
    'requests',
    ]

tests_require = [
    'freezegun',
    ]

setup(name='satosacontrib.perun',
      # ...
      install_requires=install_requires,
      tests_require=tests_require,
```

But if it was a regular dependency, it would go into `install_requires` in `setup.py`

xpavlic commented on satosacontrib/perun/micro_services/perun_user_microservice.py:
```diff
@@ -0,0 +1,142 @@
+        ]
+        self.__proxy_extsource_name = config["proxy_extsource_name"]
+        self.__allowed_requesters = config.get("allowed_requesters", [])
+        self.__registration_page_url = config.get("registration_page", [])
```
```suggestion
        self.__registration_page_url = config.get("registration_page_url", [])
```

xpavlic commented on satosacontrib/perun/config_templates/perun_user.yaml:
```diff
@@ -0,0 +1,11 @@
+  proxy_extsource_name: https://proxy.aai.muni.cz/SAML2/proxy.xml
+  allowed_requesters:
+    - https://cloud4.perun-aai.org/sp/shibboleth
+  registration_url: https://example.com/register/
```
```suggestion
  registration_page_url: https://example.com/register/
```

xpavlic reviewed (changes requested):
Rebase then last thing to change is that AdaptersManager now throws AdaptersManagerException when cannot get entity from adapters.
